pwasm-emscripten/src/lib.rs

```diff
@@ -0,0 +1,263 @@
+				Ok(Some(total_memory))
+			}
+			_ => Err(Error::Trap("not implemented".into()).into()),
+		}
+	}
+}
+
+pub fn env_module(params: EmscriptenParams) -> Result<Arc<ModuleInstanceInterface>, Error> {
+	debug_assert!(params.total_stack < params.total_memory);
+	debug_assert!((params.total_stack % LINEAR_MEMORY_PAGE_SIZE) == 0);
+	debug_assert!((params.total_memory % LINEAR_MEMORY_PAGE_SIZE) == 0);
+
+	let stack_top = params.static_size.unwrap_or(DEFAULT_STACK_TOP);
+
+	// Build module with defined memory and table,
+	// instantiate it and wrap into an Arc.
+	let instance = {
+		let builder = module()
+			// memory regions
+			.memory()
+				.with_min(params.total_memory / LINEAR_MEMORY_PAGE_SIZE)
+				.with_max(params.max_memory().map(|m| m / LINEAR_MEMORY_PAGE_SIZE))
+				.build()
+				.with_export(ExportEntry::new("memory".into(), Internal::Memory(INDEX_MEMORY)))
+			// tables
+			.table()
+				.with_min(params.table_size)
+				.build()
+				.with_export(ExportEntry::new("table".into(), Internal::Table(INDEX_TABLE)));
+		let mut instance = ModuleInstance::new(Weak::default(), "env".into(), builder.build())?;
+		instance.instantiate(None)?;
+		Arc::new(instance)
+	};
+	let total_mem_global = Arc::new(
+		VariableInstance::new(
+			false,
+			VariableType::I32,
+			RuntimeValue::I32(params.total_memory as i32),
+		).unwrap(),
+	);
+
+	let function_executor = EmscriptenFunctionExecutor {
+		total_mem_global: Arc::clone(&total_mem_global),
+	};
+
+	const SIGNATURES: &'static [UserFunctionDescriptor] = &[
+		UserFunctionDescriptor::Static("_abort", &[], None),
+		UserFunctionDescriptor::Static("abort", &[], None),
+		UserFunctionDescriptor::Static("assert", &[ValueType::I32], None),
+		UserFunctionDescriptor::Static("enlargeMemory", &[], Some(ValueType::I32)),
+		UserFunctionDescriptor::Static("getTotalMemory", &[], Some(ValueType::I32)),
+	];
+
+	let elements = UserDefinedElements {
+		executor: Some(function_executor),
+		globals: vec![
+			(
+				"STACK_BASE".into(),
+				Arc::new(
+					VariableInstance::new(
+						false,
+						VariableType::I32,
+						RuntimeValue::I32(stack_top as i32),
+					).unwrap(),
+				),
+			),
+			(
+				"STACKTOP".into(),
+				Arc::new(
+					VariableInstance::new(
+						false,
+						VariableType::I32,
+						RuntimeValue::I32(stack_top as i32),
+					).unwrap(),
+				),
+			),
+			(
+				"STACK_MAX".into(),
+				Arc::new(
+					VariableInstance::new(
+						false,
+						VariableType::I32,
+						RuntimeValue::I32((stack_top + params.total_stack) as i32),
+					).unwrap(),
+				),
+			),
+			(
+				"DYNAMIC_BASE".into(),
+				Arc::new(
+					VariableInstance::new(
+						false,
+						VariableType::I32,
+						RuntimeValue::I32((stack_top + params.total_stack) as i32),
+					).unwrap(),
+				),
+			),
+			(
+				"DYNAMICTOP_PTR".into(),
+				Arc::new(
+					VariableInstance::new(
+						false,
+						VariableType::I32,
+						RuntimeValue::I32((stack_top + params.total_stack) as i32),
```

Comment:
would require changes in parity runtime
since parity starts allocating from stack_top + total_stack and first allocation will zero-out DYNAMICTOP